### ISSUE TITLE
How to INSERT nested arrays [ {field: "one", then: "two", maybe: "three"} ]

### ISSUE THREAD

**Issue by MaxOrelus** [ClickHouse/ClickHouse#43634]
**TLDR:** I'm trying to insert records in that have an array of nested objects.

```json
{
	"name": "service",
	"description": "this is my service",
        "date": 9349447939,
	"in": [{
		"host": "localhost",
		"port": 5000
	}],
	"out": [{
		"host": "localhost",
		"port": 3000
	}]
}
```

`in` and `out` are my array of nested objects.

```sql
CREATE TABLE myTable (
  name String,
  in Nested(host String, port Int64),
  out Nested(host String, port Int64)
) ENGINE = MergeTree PARTITION BY date
ORDER BY
  date
```

```js
var db = require("....clickhouse client")

db.insert({
  table: `default.sample`,
  values: [{
    "name": "service",
    "description": "this is my service",
    "date": 9349447939,
    "in": [{
      "host": "localhost",
      "port": 5000
    }],
    "out": [{
      "host": "localhost",
      "port": 3000
    }]
  }],
  format: "JSONEachRow"
})
```

**ERROR**
`ClickHouseError: Array does not start with '[' character: (while reading the value of key in)`

Am I missing something here? Just to be clear, I'm not looking for Array of Arrays, but Array of Nested Object which can have more than +2 attributes.

**Comment by den-crane** [comment]
https://clickhouse.com/docs/en/guides/developer/working-with-json

https://kb.altinity.com/altinity-kb-schema-design/altinity-kb-jsonasstring-and-mat.-view-as-json-parser/

https://kb.altinity.com/altinity-kb-queries-and-syntax/jsonextract-to-parse-many-attributes-at-a-time/

**Comment by den-crane** [comment]
It's JSONEachRow. 

JSONEachRow is a subset of JSON 
* single line (no new line characters
* no nesting

**Comment by Avogar** [comment]
You need to change type `Array(Nested(host String, port Int64))` to just `Nested(host String, port Int64)` and disable setting `flatten_nested` (file data.jsonl contains your json example):
```sql
:) select * from file(data.jsonl, auto, 'name String, date DateTime, description String, in Nested(host String, port Int64)') settings flatten_nested=0

SELECT *
FROM file(data.jsonl, auto, 'name String, date DateTime, description String, in Nested(host String, port Int64)')
SETTINGS flatten_nested = 0

Query id: 8224cc8c-13a8-4003-a925-b3ecaf651099

┌─name────┬────────────────date─┬─description────────┬─in───────────────────┐
│ service │ 1994-01-25 15:55:47 │ this is my service │ [('localhost',5000)] │
└─────────┴─────────────────────┴────────────────────┴──────────────────────┘
```

**Comment by MaxOrelus** [comment]
Hi @Avogar @den-crane

I've tried both responses but it's still not working. I'm using the official Clickhouse library.

[https://clickhouse.com/docs/en/integrations/language-clients/nodejs/#insert-method](https://clickhouse.com/docs/en/integrations/language-clients/nodejs/#insert-method)

```js
var records = [{
  "name": "service",
  "in": [{
    "host": "localhost",
    "port": 5000
  }],
  "out": [{
    "host": "localhost",
    "port": 3000
  }]
}];
```

The object above is a **Javascript** object, not JSON at this point, but a javascript object. Can be JSON through `JSON.stringify()`

With that being said, I'm using the built-in methods provided by Clickhouse

```js
var records = [{
  "name": "service",
  "in": [{
    "host": "localhost",
    "port": 5000
  }],
  "out": [{
    "host": "localhost",
    "port": 3000
  }]
}];

(async () => {
  await client.query({
    table: "myTable",
    values: records,
    format: "JSONEachRow",
    clickhouse_settings: { flatten_nested: 0 }
  })
})()
```

I've also added the Clickhouse settings from your response, but I'm still having issues inserting the record. Could you provide examples that use the same Clickhouse lib instead of the RAW SQL queries.

I'm confused with all the JSON formats, the documentation doesn't really get me any further than where I am at this point.

**Comment by Avogar** [comment]
> I've also added the Clickhouse settings from your response, but I'm still having issues inserting the record.

This setting should be set while creating table is created, not while insert query

**Comment by alexey-milovidov** [comment]
Solved in 23.9: now we allow inserting arbitrarily nested JSON.
